Documentation/design/initphase.md

```diff
@@ -0,0 +1,44 @@
+# Init Phase
+
+## Goals
+
+1. Generates `InstallConfig` by prompting user for required parameters, and populating the rest of configuration with default values.
+2. Make useful decisions on default values for most options.
+3. Init phase should never prompt the user for more than 10 options unless there is a very strong reason.
+
+## Non Goals
+
+1. Supporting all the configuration option through init is not a goal. Init is just a helper for humans.
+
+## Overview
+
+The output of the `init` command is a config file that describes a default cluster for the given provider, and is valid/immediately usable.
+
+The `init` step is not required, this is merely a helper to generate a valid `InstallConfig`. A user could use examples from documentation, roll their own by hand, or re-use existing configs from source control.
+
+## Detailed Design
+
+### Prompts on CLI vs flags vs Environment variables
+
+1. Flags don’t work well here because we want the fast-path to be single command.
+2. Init is only valuable for humans and environment variables are similarly tedious.
+
+Init phase should use prompts on CLI for top-level configurations that are required to generate `InstallConfig`.
+
+### Remember user's answers
+
+In case where users runs `installer init` and stops after platform prompt for some reason. Should the init command ask for platform again when it is rerun.
+
+The number of prompts from users need to be very few and remembering user's answer is unnecessary optimization.
+
+### Idempotency
+
+Re-running the `installer init` prompts the user for options again and writes the `install-config.yaml` to disk, overwriting the old `install-config.yaml` file in the directory.
```

Comment:
If we allow overwrites it might make re-use of existing assets a weird UX. For example, you've run some amount of init/render/prepare/launch (so you have generated assets). Then you re run `init` with new selections generating a new `installConfig` -- this now is a mis-match to the rendered assets. Do we detect that? Do we just assume the user knows?

I don't have strong feelings here, I just originally thought saying that it would fail if it already exists punts the UX problem as it's not a supported operation.

Reply:
> If we allow overwrites it might make re-use of existing assets a weird UX...

I hadn't thought of that when [floating `--force`][1].  As long as the error message explains the constraints, I'm fine eigher way.  But I thing at some point we'll want an official workflow for folks who realized they goofed an `init` option to recover (even if it's "blow away all configs/assets/machines/... and start over").

[1]: https://github.com/openshift/installer/pull/47#discussion_r201809489

Reply:
All the steps (except `launch`) are idempotent in same sense that they re-run their responsibility overwriting previous files. If user specifically runs re-runs `init` I think the intention is to create install-config.yaml file.

Reply:
>All the steps (except launch) are idempotent in same sense that they re-run their responsibility overwriting previous files.

I'd be worried about:

1. Sometime after the first `init`, asset 123 is created.
2. Choices for a later `init` no longer require asset 123, but it gets copied out to a host anyway (or something) because it's still in an asset directory.

If later `init` runs clobber `install-config.yaml` and unwind any work based on the clobbered config (empting previos asset directories, etc.), then I think we'll be fine.